### PR TITLE
#39 - Created GET topics endpoint.

### PR DESCRIPTION
**Pull request by edwinlopez**

### CODE REVIEW

**Comment by haxwell** [comment]
w00t!

Looks good, Edwin! Give me a unit test, and we can get this committed!

**Comment by haxwell** [comment]
@jdisser Edwin's work is right along the same lines as yours with Pageable Questions, etc... See TopicRepository, especially line 17!

File: src/test/java/com/haxwell/apps/quizki/services/TopicServiceImplTest.java [hunk 1]
```diff
@@ -98,6 +112,56 @@ public void test_createNew_whenAListOfTopicsThatDoesNotAlreadyExistIsPassedIn()
...
+		list.add(new Topic(TEXT_TOPIC_1));
+		list.add(new Topic(TEXT_TOPIC_2));
+
+		Topic fullTopic1 = new Topic(TEXT_TOPIC_1);
+		fullTopic1.setId(ID_TOPIC_1);
+
+
+		when(
+			topicRepository.findByText(q, pageRequest))
+		.thenReturn(list);
```
Comment (jdisser) [inline src/test/java/com/haxwell/apps/quizki/services/TopicServiceImplTest.java:136]: Would the repository return a `Page` here not a `List`?? Also It's possible to use the `.thenAnswer()` method here to access the parameters so you can respond with a `Page` selected from a larger list of `Topic` too.

File: src/main/java/com/haxwell/apps/quizki/repositories/TopicRepository.java [hunk 2]
```diff
@@ -1,9 +1,20 @@
...
 
 import com.haxwell.apps.quizki.entities.Topic;
 
-public interface TopicRepository extends CrudRepository<Topic, Long> {
+public interface TopicRepository extends PagingAndSortingRepository<Topic, Long> {
+
 	Topic findByText(String txt);
+
+	@Query(value = "SELECT t FROM Topic t where t.text like %?1% ")
+	List<Topic> findByText(String word, Pageable pageable);
```
Comment (jdisser) [inline src/main/java/com/haxwell/apps/quizki/repositories/TopicRepository.java:17]: I think the `findByText(Pageable p)` of the `PagingAndSortingRepository` returns a `Page` here???

File: src/test/java/com/haxwell/apps/quizki/services/TopicServiceImplTest.java [hunk 3]
```diff
@@ -98,6 +112,56 @@ public void test_createNew_whenAListOfTopicsThatDoesNotAlreadyExistIsPassedIn()
...
+		int page = 1;
+		int size = 10;
+		Page<Topic> pageTopic = mock(Page.class);
+		Pageable pageRequest = PageRequest.of(page,size);
+
+		when(
+			topicRepository.findAll(pageRequest))
+		.thenReturn(pageTopic);
+
+		List<Topic> listResponse = topicService.getTopicByText("", page, size);
```
Comment (jdisser) [inline src/test/java/com/haxwell/apps/quizki/services/TopicServiceImplTest.java:158]: This doesn't seem right, the `when()` returns a `Page` but `listResponse` is a `List<Topic>` type.

**Review by jdisser — COMMENTED** [review]
I left some comments regarding using `Page` types in the object returned by the repository. For `Questions` I constructed pages based on the Pageable parameter passed in the `findBy...` methods. I was able to access them using the `.thenAnswer()` method in the `when()` chain. There's an example of this in this PR [https://github.com/team-quizki/quizki-backend/pull/47](url)

**Comment by haxwell** [comment]
@jdisser Sync the latest quizki-backend, update the database with the `src/main/resources/META-INF/sql/populate_quizki2_db_20190201.sql` file that I added, and try Edwin's change.. :)

Spoiler: It works.

**Comment by jdisser** [comment]
I figured it did! I was not certain how it does. But my mistake was thinking of `List` as a type instead of an interface and the same with `Page`. Both are children of `Iterable` and so it flys!

**Comment by jdisser** [comment]
A better explanation as to why it works is given here
[https://docs.spring.io/spring-data/data-commons/docs/1.6.1.RELEASE/reference/html/repositories.html](url)
**Section 1.2.2 Special Parameter Handling**

An interesting feature is that the returned list can be ordered too! Just put an `OrderByAsc` in the method name. Might be useful.

**Pull request merged**